### PR TITLE
feat(common): improve builder parameter passing for child and dep builds

### LINKED ISSUE #11394: bug(common): children of dependencies in builder scripts do not inherit dependency build mode
Given we have the following when we do a build of Developer Server:

```
[developer/src/server] child build, parameters: <build --builder-child --deps>
[developer/src/server] ## build:server starting...
[web] dependency build, started by developer/src/server
[web] build.sh parameters: <configure build --deps --builder-dep-parent developer/src/server>
[web/src/app/browser] child build, parameters: <configure --builder-child --deps>
[web/src/app/browser] ## configure starting...
[web/src/app/browser] Skipping dependency web/src/engine/attachment for configure
[web/src/app/browser] Skipping dependency web/src/engine/main for configure
[web/src/tools/building/sourcemap-root] dependency build, started by web/src/app/browser
[web/src/tools/building/sourcemap-root] build.sh parameters: <configure build --deps --builder-dep-parent web/src/app/browser>
[web/src/tools/building/sourcemap-root] skipping configure, up-to-date
[web/src/tools/building/sourcemap-root] skipping build, up-to-date
npm WARN skipping integrity check for git dependency ssh://git@github.com/keymanapp/dependency-node-xml2js.git 
npm WARN deprecated @npmcli/move-file@2.0.1: This functionality has been moved to @npmcli/fs
...
```

web/src/app/browser runs as a child build rather than as a dependency build, losing `--builder-dep-parent`. This is probably backwards, as it means we do unnecessary builds when trying to build deps.

### LINKED ISSUE #11408: feat(common): if `build.sh` is called with `--builder-child`, then hint on actions and options that are not specified, rather than failing
Pros:
 - less noise in build script sources for unused params

Cons:
 - could miss important parameters that are not defined for given child script

For example, `publish-sentry` action missing for a given module would mean that no symbols would be published to sentry for that module, but this would not result in an error if called from a higher-level script. It's hard to see a way to address this, perhaps a validation report would do the job?

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #11408.
Fixes #11394.

@keymanapp-test-bot skip

### CODE REVIEW

File: resources/builder.inc.sh [hunk 1]
```diff
@@ -1369,17 +1378,26 @@ _builder_parse_expanded_parameters() {
           # internal use parameter for dependency builds - identifier of parent script
           shift
           builder_dep_parent="$1"
+          builder_echo setmark "dependency build, started by $builder_dep_parent"
+          builder_echo grey "build.sh parameters: <${_params[@]}>"
```
Comment (mcdurdin) [inline resources/builder.inc.sh:1382]: Moves the reporting ahead of any other messages that might appear, see also ll.1386, for example debug messages on ll.1397 should appear after the note saying that we are a child build or dep build.

File: resources/builder.inc.sh [hunk 2]
```diff
@@ -400,9 +400,19 @@ _builder_execute_child() {
...
+  if builder_is_dep_build; then
+    dep_flag=--builder-dep-parent
+    dep_module="$builder_dep_parent"
+  fi
+
+  "$script" \
     --builder-child \
     $_builder_build_deps \
+    $dep_flag "$dep_module" \
+    $action \
```
Comment (mcdurdin) [inline resources/builder.inc.sh:415]: The action has to come after the --builder-child flag so that we can ignore if we are in --builder-child mode.

**Comment by darcywong00** [comment]
Does any of this need to be clarified in the doc?
https://github.com/keymanapp/keyman/blob/master/resources/build/builder.md#defining-build-script-parameters

File: resources/build/test/ignored-flags/child/build.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+
+echo "Params: $@"
```
Comment (darcywong00) [inline resources/build/test/ignored-flags/child/build.sh:3]: Is this for debugging?

**Review by jahorton — APPROVED** [review]
For the `debug-deps` test, it would be nice to have an action that verifies that a child action is skipped if it was already run due to dependency mode.

So, a setup like...

parent:
```sh
builder_describe "foo" \
"@/resources/build/test/debug-deps/dep/child" \  
"@/resources/build/test/debug-deps/dep"
```

Of course, a more likely case is where the first 'dependency' there is triggered by a different dependency that happens to depend on `dep/child`.

----

Aside from that, I think this looks fine.

File: resources/build/test/ignored-flags/child/build.sh (hunk 3 above)
Comment (mcdurdin) [reply resources/build/test/ignored-flags/child/build.sh:3]: good catch

**Comment by mcdurdin** [comment]
> Does any of this need to be clarified in the doc?

yes, will push an update, thanks

> For the `debug-deps` test, it would be nice to have an action that verifies that a child action is skipped if it was already run due to dependency mode.

yes, will add

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.34-alpha](https://keyman.com/downloads/releases/alpha/18.0.34)

**Pull request merged**